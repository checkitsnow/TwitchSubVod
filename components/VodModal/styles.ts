import styled from 'styled-components';

interface AdsProps {
  isVisible: boolean;
}

export const Container = styled.div`
  margin-top: 2rem;
  position: relative;
  width: 60rem;
  height: 33.75rem;

  .react-player {
    position: absolute;
    top: 0;
    left: 0;

    iframe {
      height: 100vh;
    }
  }

  @media (max-width: 960px) {
    width: 48rem;
    height: 27rem;
  }

  @media (max-width: 768px) {
    width: 30rem;
    height: 16.875rem;
  }

  @media (max-width: 480px) {
    width: 22.5rem;
    height: 12.65625rem;
  }
`;

export const Ads = styled.div<AdsProps>`
  z-index: ${({ isVisible }) => (isVisible ? '10' : '-1')};
  padding: 10% 20%;
  display: flex;
  flex-direction: column;
  position: absolute;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.7);
  overflow: hidden;
  justify-content: center;
  align-items: center;

  a {
    color: #fafafa;
    background: #323b43;
    padding: 3rem 2rem;
    border-radius: 8px;
    transition: background 0.2s ease;

    p {
      margin: 0;
      opacity: 0.4;
      width: 100%;
      text-align: end;
    }

    :hover {
      background: #1b2125;
    }
  }
  @media (max-width: 768px) {
    padding: 0;

    a {
      padding: 2rem 2rem;
      margin-top: 0;
    }
  }

  button {
    top: 5%;
    right: 3%;
    position: absolute;
    padding: 0.1rem 0.3rem;
    border-radius: 5px;
    line-height: 0;
    background: var(--purple);
    color: inherit;
    font-weight: bold;
    display: flex;
    align-items: center;
    transition: background 0.2s ease;

    svg {
      margin-left: 0.5rem;
    }

    :hover {
      background: var(--dark-purple);
    }
  }
`;
