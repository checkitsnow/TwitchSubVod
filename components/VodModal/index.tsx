import React, { useRef, useEffect, useMemo, useState } from 'react';
import ReactGA from 'react-ga';
import ReactPlayer from 'react-player';
import { FaWindowClose } from 'react-icons/fa';
import { Player } from 'video-react';

import { useGlobal } from '@/stores/GlobalContext';
import { Container, Ads, CustomOptions } from './styles';

// 7days * 24 hours * 60 minutes * 60 seconds * 1000ms
const timeInterval = 7 * 24 * 60 * 60 * 1000;

const VodModal = ({ videoUrl, previewUrl }: any) => {
  const [playbackRate, setPlaybackRate] = useState(1);
  const playerRef: any = useRef(null);

  const { videoQuality } = useGlobal();
  const [showAd, setShowAd] = useState(false);
  useEffect(() => {
    const paymentTime = Number(localStorage.getItem('paymentTime'));

    if (Number(Date.now()) - paymentTime > timeInterval) {
      setShowAd(true);
    }
  }, [videoUrl]);

  const handleClick = () => {
    localStorage.setItem('paymentTime', Date.now().toString());
    setShowAd(false);
    ReactGA.event({
      category: 'AlreadyPaid',
      action: videoUrl.split('_')[1],
    });
  };

  const downloadClip = () => {
    const currentTime = playerRef?.current?.getCurrentTime() || 0;

    const clippedPart = Math.floor(currentTime / 10);

    const clippedUrl = videoUrl
      .replace('index-dvr.m3u8', `${clippedPart}.ts`)
      .replace(process.env.NEXT_PUBLIC_CORS, '');

    const link = document.createElement('a');
    link.href = clippedUrl;
    link.setAttribute('download', 'clip');
    document.body.appendChild(link);
    link.click();
  };

  const showCustomPlayer =
    /^iP/.test(navigator.platform) ||
    (/^Mac/.test(navigator.platform) && navigator.maxTouchPoints > 4);

  const renderVodModal = useMemo(() => {
    return (
      <>
        <Ads isVisible={showAd}>
          <div>
            <a
              href="https://ko-fi.com/pogulive"
              target="_blank"
              rel="noopener noreferrer"
            >
              Hey, these servers aren't free...
              <img
                className="kofi-img"
                width="230px"
                src="https://cdn.ko-fi.com/cdn/kofi5.png?v=2"
                alt="Buy Me a Coffee at ko-fi.com"
              />
              <p>{`... if you can, of course :)`}</p>
            </a>
            <button onClick={handleClick}>
              I already paid
              <img
                title="EleGiggle"
                src="https://static-cdn.jtvnw.net/emoticons/v1/4339/3.0"
              />
            </button>
          </div>
          <button onClick={() => setShowAd(false)}>
            {window.innerWidth > 480 && 'Close Reminder'}
            <FaWindowClose size={32} />
          </button>
        </Ads>

        {showCustomPlayer ? (
          <>
            <Player
              playsInline
              poster={previewUrl}
              src={videoUrl.replace(process.env.NEXT_PUBLIC_CORS, '')}
            />
          </>
        ) : (
          <>
            <CustomOptions>
              <select
                defaultValue="1"
                onChange={(e) => setPlaybackRate(Number(e.target.value))}
              >
                <option value="0.5">Speed 0.5x</option>
                <option value="0.75">Speed 0.75x</option>
                <option value="1">Speed 1x</option>
                <option value="2">Speed 2x</option>
              </select>

              <button onClick={downloadClip}>
                Clip it{' '}
                <img
                  title="LUL"
                  src="https://static-cdn.jtvnw.net/emoticons/v2/425618/default/dark/3.0"
                />
              </button>
            </CustomOptions>

            <ReactPlayer
              ref={playerRef}
              url={videoUrl}
              controls
              width="100%"
              height="100%"
              playbackRate={playbackRate}
              config={{
                file: {
                  hlsOptions: {
                    xhrSetup: (xhr: any, _url: string) => {
                      xhr.open(
                        'GET',
                        _url
                          .replace('unmuted.ts', 'muted.ts')
                          .replace('chunked', videoQuality),
                        true,
                      );
                    },
                  },
                },
              }}
            />
          </>
        )}
      </>
    );
  }, [
    videoUrl,
    showAd,
    setShowAd,
    videoQuality,
    setPlaybackRate,
    playbackRate,
    downloadClip,
  ]);

  return <Container>{renderVodModal}</Container>;
};

export default VodModal;
