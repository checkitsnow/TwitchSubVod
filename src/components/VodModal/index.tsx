import React from 'react';
import ReactPlayer from 'react-player';

import { Container } from './styles';

const VodModal = ({ videoUrl }: any) => {
  return (
    <Container>
      <ReactPlayer url={videoUrl} controls width="100%" height="100%" />
    </Container>
  );
};

export default VodModal;
