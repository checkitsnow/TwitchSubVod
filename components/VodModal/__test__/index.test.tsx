import { render } from '@testing-library/react';
import VodModal from '..';

it('should render Vod Modal correctly', () => {
  const tree = render(
    <VodModal videoUrl="https://www.youtube.com/watch?v=d1YBv2mWll0" />,
  );
  expect(tree).toMatchSnapshot();
});
