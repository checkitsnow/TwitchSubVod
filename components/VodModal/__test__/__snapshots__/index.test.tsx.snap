// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Vod Modal correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="styles__Container-xygqpp-0 hoGEtO"
      >
        <div
          class="styles__Ads-xygqpp-1 ccrZEW"
        >
          <div>
            <a
              href="https://ko-fi.com/pogulive"
              rel="noopener noreferrer"
              target="_blank"
            >
              Hey, these servers aren't free...
              <img
                alt="Buy Me a Coffee at ko-fi.com"
                class="kofi-img"
                src="https://cdn.ko-fi.com/cdn/kofi5.png?v=2"
                width="230px"
              />
              <p>
                ... if you can, of course :)
              </p>
            </a>
            <button>
              I already paid
              <img
                src="https://static-cdn.jtvnw.net/emoticons/v1/4339/3.0"
                title="EleGiggle"
              />
            </button>
          </div>
          <button>
            Close Reminder
            <svg
              fill="currentColor"
              height="32"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z"
              />
            </svg>
          </button>
        </div>
        <div
          class="styles__CustomOptions-xygqpp-2 kMIVgN"
        >
          <select>
            <option
              value="0.5"
            >
              Speed 0.5x
            </option>
            <option
              value="0.75"
            >
              Speed 0.75x
            </option>
            <option
              selected=""
              value="1"
            >
              Speed 1x
            </option>
            <option
              value="2"
            >
              Speed 2x
            </option>
          </select>
          <button>
            Clip it
             
            <img
              src="https://static-cdn.jtvnw.net/emoticons/v2/425618/default/dark/3.0"
              title="LUL"
            />
          </button>
        </div>
        <div
          style="width: 100%; height: 100%;"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="styles__Container-xygqpp-0 hoGEtO"
    >
      <div
        class="styles__Ads-xygqpp-1 ccrZEW"
      >
        <div>
          <a
            href="https://ko-fi.com/pogulive"
            rel="noopener noreferrer"
            target="_blank"
          >
            Hey, these servers aren't free...
            <img
              alt="Buy Me a Coffee at ko-fi.com"
              class="kofi-img"
              src="https://cdn.ko-fi.com/cdn/kofi5.png?v=2"
              width="230px"
            />
            <p>
              ... if you can, of course :)
            </p>
          </a>
          <button>
            I already paid
            <img
              src="https://static-cdn.jtvnw.net/emoticons/v1/4339/3.0"
              title="EleGiggle"
            />
          </button>
        </div>
        <button>
          Close Reminder
          <svg
            fill="currentColor"
            height="32"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z"
            />
          </svg>
        </button>
      </div>
      <div
        class="styles__CustomOptions-xygqpp-2 kMIVgN"
      >
        <select>
          <option
            value="0.5"
          >
            Speed 0.5x
          </option>
          <option
            value="0.75"
          >
            Speed 0.75x
          </option>
          <option
            selected=""
            value="1"
          >
            Speed 1x
          </option>
          <option
            value="2"
          >
            Speed 2x
          </option>
        </select>
        <button>
          Clip it
           
          <img
            src="https://static-cdn.jtvnw.net/emoticons/v2/425618/default/dark/3.0"
            title="LUL"
          />
        </button>
      </div>
      <div
        style="width: 100%; height: 100%;"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
